// @simplewebauthn/server/script/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.verifyRegistrationResponse = exports.verifyAuthenticationResponse = exports.SettingsService = exports.MetadataService = exports.generateRegistrationOptions = exports.generateAuthenticationOptions = void 0;
/**
 * @packageDocumentation
 * @module @simplewebauthn/server
 */
const generateRegistrationOptions_js_1 = require("./registration/generateRegistrationOptions.js");
Object.defineProperty(exports, "generateRegistrationOptions", { enumerable: true, get: function () { return generateRegistrationOptions_js_1.generateRegistrationOptions; } });
const verifyRegistrationResponse_js_1 = require("./registration/verifyRegistrationResponse.js");
Object.defineProperty(exports, "verifyRegistrationResponse", { enumerable: true, get: function () { return verifyRegistrationResponse_js_1.verifyRegistrationResponse; } });
const generateAuthenticationOptions_js_1 = require("./authentication/generateAuthenticationOptions.js");
Object.defineProperty(exports, "generateAuthenticationOptions", { enumerable: true, get: function () { return generateAuthenticationOptions_js_1.generateAuthenticationOptions; } });
const verifyAuthenticationResponse_js_1 = require("./authentication/verifyAuthenticationResponse.js");
Object.defineProperty(exports, "verifyAuthenticationResponse", { enumerable: true, get: function () { return verifyAuthenticationResponse_js_1.verifyAuthenticationResponse; } });
const metadataService_js_1 = require("./services/metadataService.js");
Object.defineProperty(exports, "MetadataService", { enumerable: true, get: function () { return metadataService_js_1.MetadataService; } });
const settingsService_js_1 = require("./services/settingsService.js");
Object.defineProperty(exports, "SettingsService", { enumerable: true, get: function () { return settingsService_js_1.SettingsService; } });
